### PR TITLE
types: support sql_mode ALLOW_INVALID_DATES

### LINKED ISSUE #8263: sql_mode ALLOW_INVALID_DATES is not supported
## Feature Request

**Is your feature request related to a problem? Please describe:**

Relates to #2738 - currently `ALLOW_INVALID_DATES` is not supported.  It is not commonly used, but could be useful for compatibility.

**Describe the feature you'd like:**
```
# testcase
CREATE TABLE d1 (d DATE);
SET sql_mode='STRICT_TRANS_TABLES,ALLOW_INVALID_DATES';
INSERT INTO d1 VALUES ('2010-00-01');
```
Passes in MySQL, produces this in TiDB:
```
MySQL [test]> SET sql_mode='STRICT_TRANS_TABLES,ALLOW_INVALID_DATES';
ERROR 1105 (HY000): ERROR 1231 (42000): Variable 'sql_mode' can't be set to the value of 'ALLOW_INVALID_DATES'
MySQL [test]> INSERT INTO d1 VALUES ('2010-00-01');
ERROR 1292 (22007): Incorrect datetime value: '2010-00-01'
```

**Describe alternatives you've considered:**
Possible to document as unsupported.

**Teachability, Documentation, Adoption, Migration Strategy:**
Improves compatibility with MySQL.

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve? 
support mode `ALLOW_INVALID_DATES` (fix #8263). With this mode enabled it will not perform full checking of dates. Check only that the month is in the range from 1 to 12 and the day is in the range from 1 to 31 (also allow zero values such as `0000-00-00`).

### What is changed and how it works?
Add a new flag `AllowInvalidDate` into `StatementContext`. Then `checkMonthDay` check the month and day according to this flag.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

File: executor/insert_test.go [hunk 1]
```diff
@@ -14,6 +14,7 @@
 package executor_test
 
 import (
+	"fmt"
```
Comment (jackysp) [inline executor/insert_test.go]: It is better to separate the sys lib with the 3rd ones.

File: executor/insert_test.go [hunk 2]
```diff
@@ -233,3 +235,34 @@ func (s *testSuite3) TestInsertZeroYear(c *C) {
...
+	tk.MustExec(`drop table if exists t1;`)
+	tk.MustExec(`create table t1(d date);`)
+	tk.MustExec(`create table t2(d datetime);`)
+	tk.MustExec(`create table t3(d date);`)
+	tk.MustExec(`create table t4(d datetime);`)
+
+	runWithMode := func(mode string) {
+		inputs := []string{"0000-00-00", "2019-00-00", "2019-01-00", "2019-00-01", "2019-02-31"}
+		results := testkit.Rows(`0 0 0`, `2019 0 0`, `2019 1 0`, `2019 0 1`, `2019 2 31`)
+
```
Comment (imtbkcat) [inline executor/insert_test.go]: Maybe we should set the original sql_mode back after finishing this test, using defer in line 250.

**Comment by bobotu** [comment]
PTAL @imtbkcat

**Pull request merged**